core/utils/codel_test.cc

```diff
@@ -0,0 +1,234 @@
+#include "./codel.h"
+
+#include <gtest/gtest.h>
+#include <math.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "../mem_alloc.h"
+#include "../packet.h"
+#include "./lock_free_queue.h"
+#include "./time.h"
+
+namespace {
+
+using bess::utils::Codel;
+using bess::utils::LockFreeQueue;
+void integer_drop(int* ptr) {
+  delete ptr;
+}
+
+// Simple test to make sure one can get back out pointer
+TEST(CodelTest, SingleInputOutput) {
+  auto q = new LockFreeQueue<Codel<int>::Wrapper>(8);
+  Codel<int> c(&integer_drop, q, 5000, 100000);
+  int* val = new int();
+  ASSERT_NE(val, nullptr);
+  ASSERT_FALSE(c.Push(val));
+  ASSERT_EQ(c.Size(), 1);
+  int* output = c.Pop();
+  ASSERT_EQ(output, val);
+  ASSERT_EQ(c.Size(), 0);
+  EXPECT_EQ(output, val);
+  delete val;
+}
+
+// Tests to make sure that low delay entries won't be dropped
+TEST(CodelTest, NoDropTest) {
+  auto q = new LockFreeQueue<Codel<int>::Wrapper>(8);
+  Codel<int> c(&integer_drop, q, 5000, 100000);
+
+  int n = 4;
+  int* vals[n];
+  for (int i = 0; i < n; i++) {
+    vals[i] = new int();
+  }
+  for (int i = 0; i < n; i++) {
+    ASSERT_FALSE(c.Push(vals[i]));
+  }
+  EXPECT_EQ(c.Size(), n);
+
+  usleep(3000);  // low delay should not cause any drops
+  int* output = c.Pop();
+  ASSERT_EQ(output, vals[0]);
+  output = c.Pop();
+  ASSERT_EQ(output, vals[1]);
+  output = c.Pop();
+  ASSERT_EQ(output, vals[2]);
+  output = c.Pop();
+  ASSERT_EQ(output, vals[3]);
+  EXPECT_EQ(c.Size(), 0);
+
+  for (int i = 0; i < n; i++) {
+    delete vals[i];
+  }
+}
+
+// Tests to make sure that codel enters drop state at the proper time and
+// properly drops at the correct time.
+TEST(CodelTest, DropTest) {
+  auto q = new LockFreeQueue<Codel<int>::Wrapper>(16);
+  Codel<int> c(&integer_drop, q, 5000, 100000);
+
+  int n = 10;
+  int* vals[n];
+  for (int i = 0; i < n; i++) {
+    vals[i] = new int();
+  }
+  for (int i = 0; i < n; i++) {
+    ASSERT_FALSE(c.Push(vals[i]));
+  }
+  EXPECT_EQ(c.Size(), n);
+
+  usleep(10000);  // all packets will have greater delay than target of 5ms
+  int* output;
+
+  // should mark above target after dequeuing this entry
+  output = c.Pop();
+  ASSERT_EQ(output, vals[0]);
+
+  // still shouldn't drop yet because hasn't been interval since
+  // being above target
+  output = c.Pop();
+  ASSERT_EQ(output, vals[1]);
+  usleep(105000);  // wait above a window of time to make next entry drop
+
+  // Should be above target for window.
+  // should drop vals[2], enter drop state, and return 4th instead.
+  output = c.Pop();
+  ASSERT_EQ(output, vals[3]);
+
+  // should now in be drop state and the next drop should be 100ms.
+  usleep(72000);
+  output = c.Pop();
+  ASSERT_EQ(output, vals[4]);  // shouldn't be a drop yet only has been 72ms.
+
+  usleep(30000);
+  output = c.Pop();
+  ASSERT_EQ(output,
```

Comment:
For weird line breaks like this, move the comment before the line and merge the code line.